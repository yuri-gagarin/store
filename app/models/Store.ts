import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;