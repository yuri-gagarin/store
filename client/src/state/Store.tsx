import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

