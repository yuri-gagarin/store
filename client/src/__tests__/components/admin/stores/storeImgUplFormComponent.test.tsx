import React from "react";
import { Button } from "semantic-ui-react";
import moxios from "moxios";
// testing imports //
import { mount, shallow, ReactWrapper, ShallowWrapper } from "enzyme";
import { act } from 'react-dom/test-utils';
// components //
import StoreImgUplForm from "../../../../components/admin_components/stores/forms/StoreImageUplForm";
import { StateProvider } from "../../../../state/Store";
// helpers //
import MockFile from "../../../helpers/mockFile";
import { createMockStores } from "../../../../test_helpers/storeHelpers";

describe("Store Image Upload Form Tests", () => {

  describe("Render tests without any Image data", () => {
    let component: ShallowWrapper<React.FC>;

    beforeAll(() => {
      component = shallow<React.FC<{}>, {}>(<StoreImgUplForm />)
     });

    it("Should properly render", () => {
      expect(component).toMatchSnapshot();
    });
    it("Should have one upload Button element", () => {
      const uplButton = component.find(Button);
      expect(uplButton.length).toEqual(1);
    });
    it("Should have an input", () => {
      const input = component.find("input");
      expect(input.length).toEqual(1);
    });
  });
  
  describe("Render tests with an Image file present", () => {
    let component: ShallowWrapper<React.FC>;
    let input: ShallowWrapper;

    beforeAll(() => {
      component = shallow<React.FC<{}>, {}>(<StoreImgUplForm />);
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });

    it("Should properly set the file in props", () => {
      expect(input.prop("type")).toBeDefined();
      expect(input.prop("type")).toEqual("file");
    });
    it("Should successfully update and render the ImgUpload button", () => {
      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should successfully update and render the Cancel button", () => {
      const imgCanceldtn = component.find("#storeImgCancelBtn");
      expect(imgCanceldtn.length).toEqual(1);
    });
    it("Should properly handle the '#storeImgCancelBtn' and render correct local state", () => {
      const imgCancelBtn = component.find("#storeImgCancelBtn");
      imgCancelBtn.simulate("click");
      expect(component.find("#storeImgCancelBtn").length).toEqual(0);
      expect(component.find("#storeImgUploadBtn").length).toEqual(0);
      expect(component.find("#selectStoreImgBtn").length).toEqual(1);
    });
  });
  // MOCK Successful Image upload tests //
  describe("'#storeImgUploadBtn' functionality, successful upload and local state changes", () => {
    let component: ReactWrapper<typeof StoreImgUplForm, {}, {}>;
    let input: ReactWrapper;

    beforeAll(() => {
      component = mount<React.FC<{}>, {}>(
        <StateProvider>
          <StoreImgUplForm />
        </StateProvider>
      );
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    afterEach(() => {
      moxios.uninstall();
    });

    it("Should successfully fire the 'uploadStoreImg' action and show loader", async () => {
      moxios.install();
      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 200,
          response: {
            responseMsg: "All ok",
            updatedStore: createMockStores(1)[0]
          }
        });
      });

      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      imgUpoadBtn.at(0).simulate("click");
      // loader should be displayed on the #storeImgUploadBtn //
      const imgUpoadBtnLoading = component.find(Button);
      expect(imgUpoadBtnLoading.at(1).props().loading).toEqual(true);
      // update component and local state //
      await act( async () => {
        return new Promise((res, _) => {
          setTimeout(() => {
            component.update();
            res();
          }, 500);
        });
      });
      
    });
    it("Should correctly render Select Image button after 'successful upload", () => {
      const selectImgBtn = component.find("#selectStoreImgBtn");
      expect(selectImgBtn.length).toEqual(2);
    });
    it("Should NOT render Image Upload button after 'successful upload", () => {
      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
    it("Should NOT render Cancel Upload button after 'successful upload", () => {
      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
  });
  // END Mock successful Image upload tests //
  // MOCK unsuccessful Image upload tests //
  describe("'#uploadStoreImgUploadBtn' functionality and a failed upload", () => {
    let component: ReactWrapper<React.FC>;
    let input: ReactWrapper;

    beforeAll(() => {
      component = mount<React.FC<{}>, {}>(
        <StateProvider>
          <StoreImgUplForm />
        </StateProvider>
      );
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    afterEach(() => {
      moxios.uninstall();
    });

    it("Should successfully handle the 'uploadStoreImg' action error", async () => {
      moxios.install();
      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 500,
          response: {
            responseMsg: "Error",
            error: new Error("Oops")
          }
        });
      });

      const imgUpoadBtn = component.find("#storeImgUploadBtn").at(0);
      imgUpoadBtn.simulate("click");
      // loader should be displayed on the #storeImgUploadBtn //
      const imgUpoadBtnLoading = component.find(Button);
      expect(imgUpoadBtnLoading.at(1).props().loading).toEqual(true);
      // update component and local state //
      await act( async () => {
        return new Promise((res, _) => {
          setTimeout(() => {
            component.update();
            res();
          }, 500);
        });
      });
    });
    it("Should NOT render Select Image button after a 'failed' upload", () => {
      const selectImgBtn = component.find("#selectStoreImgBtn");
      expect(selectImgBtn.length).toEqual(0);
    });
    it("Should render Image Upload button after a 'failed' upload", () => {
      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(2);
    });
    it("Should render Cancel Upload button after a 'failed' upload", () => {
      const imgUpoadBtn = component.find("#storeImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(2);
    });
  });
  // END Mock unsuccessful Image upload tests //
});