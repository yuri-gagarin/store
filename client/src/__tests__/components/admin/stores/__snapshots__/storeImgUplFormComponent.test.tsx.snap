// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Image Upload Form Tests Render tests without any Image data Should properly render 1`] = `
<div
  id="storeImgUplFormHolder"
>
  <div>
    <p>
      Image Uploader
    </p>
  </div>
  <Form
    as="form"
    id="storeImgUploadForm"
  >
    <div
      id="storeImgInputControlls"
    >
      <Button
        as="label"
        content="Choose Image"
        htmlFor="storeImgFile"
        icon="file"
        id="selectStoreImgBtn"
        labelPosition="left"
        onClick={[Function]}
      />
      <input
        hidden={true}
        id="storeImgFile"
        onChange={[Function]}
        type="file"
      />
    </div>
    <span>
      No File
    </span>
  </Form>
</div>
`;
